widget/completion_entry_test.go

```diff
@@ -0,0 +1,161 @@
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("init")
+	assert.True(t, entry.popupMenu.Visible())
+
+}
+
+// Navigate in menu and select an entry.
+func TestCompletionEntry_Navigate(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("init")
+
+	// navigate to "bar" and press enter, the entry should contain
+	// "bar" in value
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyReturn})
+
+	assert.Equal(t, "bar", entry.Text)
+	assert.False(t, entry.popupMenu.Visible())
+}
+
+// Hide the menu on Escape key.
+func TestCompletionEntry_Escape(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("init")
+
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyEscape})
+
+	assert.False(t, entry.popupMenu.Visible())
+}
+
+// Hide the menu on rune pressed.
+func TestCompletionEntry_Rune(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("foobar")
+	entry.CursorColumn = 6 // place the cursor after the text
+
+	// type some chars...
+	win.Canvas().Focused().TypedRune('x')
+	win.Canvas().Focused().TypedRune('y')
+	assert.Equal(t, "foobarxy", entry.Text)
+
+	// make a move and type other char
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	win.Canvas().Focused().TypedRune('z')
+	assert.Equal(t, "foobarxyz", entry.Text)
+
+	assert.True(t, entry.popupMenu.Visible())
+}
+
+// Hide the menu on rune pressed.
+func TestCompletionEntry_Rotation(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.SetText("foobar")
+	entry.CursorColumn = 6 // place the cursor after the text
+
+	// loop one time (nb items + 1) to go back to the first element
+	for i := 0; i <= len(entry.Options); i++ {
+		win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyDown})
+	}
+
+	assert.Equal(t, 0, entry.navigableList.selected)
+
+	// Do the same in reverse order, here, onlh one time to go on the last item
+	win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyUp})
+	assert.Equal(t, len(entry.Options)-1, entry.navigableList.selected)
+}
+
+// Test sumbission with opened completion.
+func TestCompletionEntry_OnSubmit(t *testing.T) {
+	entry := createEntry()
+	win := test.NewWindow(entry)
+	win.Resize(fyne.NewSize(500, 300))
+	defer win.Close()
+
+	entry.OnSubmitted = func(s string) {
+		entry.HideCompletion()
+		assert.True(t, entry.popupMenu.Hidden)
+	}
```

Comment:
I am aware this test works, but maybe it is better to ensure this is called otherwise the assertion could not be performed and the test would be a false positive:
```go
submitCalled := false
entry.OnSubmitted = func(s string) {
	submitCalled = true
	entry.HideCompletion()
	assert.True(t, entry.popupMenu.Hidden)
}
...
win.Canvas().Focused().TypedKey(&fyne.KeyEvent{Name: fyne.KeyReturn})
assert.True(t, submitCalled)
```

Reply:
I'm not OK with this, this test is made to check an issue where the "OnSumitted" should not be called. Testing if the submission is made is done in the other test.

Reply:
I am confused then, this test is called:
```go
// Test sumbission with opened completion.
func TestCompletionEntry_OnSubmit(t *testing.T) {
...
}
```
Isn't this testing the submission?

Reply:
My god, sorry I didn't seen the "submitCalled" variable you added. OK, no problem then.

Reply:
Done https://github.com/fyne-io/fyne-x/pull/11/commits/d5226ec4ef7f385c9ef6207bfa7a68a8ce37d30c

Reply:
> My god, sorry I didn't seen the "submitCalled" variable you added

No worries :)